adagrams/game.py

```diff
@@ -1,11 +1,64 @@
+from random import sample
 def draw_letters():
-    pass
+    hand = sample(['A','B','C','D','E','F','G','H','I','J','K','L',\
+                   'M','N','O','P','Q','R','S','T','U','V','W','X',\
+                   'Y','Z'], counts = [9,2,2,4,12,2,3,2,9,1,1,4,2,6,\
+                                       8,2,1,6,4,6,4,2,2,1,2,1], k = 10)
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letter_bank_dict = {}
+    for letter in letter_bank:
+        if letter not in letter_bank_dict:
+            letter_bank_dict[letter] = 1
+        else:
+            letter_bank_dict[letter] += 1
+    for letter in word.upper():
+        if letter not in letter_bank_dict:
+            return False
+        elif letter_bank_dict[letter] == 0:
+            return False
+        else:
+            letter_bank_dict[letter] -= 1
+    return True
+
 
 def score_word(word):
-    pass
+    scores = {
+        "AEIOULNRST" : 1,
+        "DG" : 2,
+        "BCMP" : 3,
+        "FHVWY" : 4,
+        "K" : 5,
+        "JX" : 8,
+        "QZ" : 10
+    }
+
+    score = [val for letter in word.upper() for key, val in scores.items() if letter in key]
+
+    if len(score) >= 7:
+        score.append(8)
+
+    return sum(score)
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    score_list = [(word, score_word(word)) for word in word_list]
+
+    highest_scoring = [score_list[0]]
+
+    for word in score_list[1:]:
+        if word[1] == highest_scoring[0][1]:
+            highest_scoring.append(word)
+        elif word[1] > highest_scoring[0][1]:
+            highest_scoring.clear()
+            highest_scoring.append(word)
+
+    if len(highest_scoring) == 1:
+        return highest_scoring[0]
+
+    highest_scoring.sort(key=lambda word: len(word[0]))
+
+    if len(highest_scoring[-1][0]) == 10:
+        highest_scoring.sort(key=lambda word: len(word[0]), reverse=True)
```

Comment:
While this `if` block will usually select a 10-letter word as a tiebreaker, and we can assume no word uses more than 10 letters, and thus the 10 letter words are at the end of `highest_scoring`, there is a logic error if there are _multiple_ 10-letter words that causes this implementation to not match the requirements.

From the project requirements: "If the there are multiple words that are the same score and the same length, pick the first one in the supplied list". Given that, with two 10-letter highest scoring words, we would want to pick whichever word was given first in `word_list`. However, y'all's code is picking the _last_ word in `highest_scoring` (if it has 10 letters), which will _not_ be the 10-letter word that came first in `word_list` but the 10-letter word that came _last_.

My suggested `for` loop in above comment avoids this error, obviously...

Reply:
You may be wondering how I knew that the last word in `highest_scoring` was also the last relevant word in `word_list` despite the call to `highest_scoring.sort` on line 59. Basically the `sort` function in Python is guaranteed to be a stable sort. That means when there are ties, the order from the original list is maintained.

So let's go through the full genealogy of `highest_scoring`. The list comprehension that sets `score_list` will maintain the order of `word_list`. We then go through `score_list` to append to `highest_scoring`, so that order will be maintained. And then since `sort` is a stable sort, even calling that will maintain the order among ties.